### ISSUE TITLE
Processing P3D windows takes 30 seconds to 1 minute to appear ...

### ISSUE THREAD

**Issue by nkolban** [processing/processing#4239]
I am new to processing and have been running it just fine on my Windows 7 laptop. I decided I wanted to run it on my deskside which has dual monitors and a high end graphics card and runs Windows 10. I installed processing without incident but when I try and run a sketch ... and I hit the play button ... the display windows takes 1-2 minutes to appear. After that, I can interact with the display perfectly. If I close the display and re-start the sketch, again ... 1-2 minutes to appear. I have rebooted my PC and tried to find any error logs ... but I can't seem to find any issues ... it is just frustratingly slow to show the display. I searched the forums but couldn't find any obvious matches. Has anyone heard/seen anything like this before? I can demonstrate live if that would help or else provide any logs that may be useful.

This issue was originally started as a forum thread and discussion on it can be found here ...

https://forum.processing.org/two/discussion/14454/p3d-not-always-creating-window

**Comment by alignedleft** [comment]
For a few weeks, Processing 3.0.1 was super slow on initial open on my Mac. But then eventually it went back to its usual, speedy self. Could have been that I applied a Java update or something else behind the scenes.

**Comment by nkolban** [comment]
Let me know if I can provide any diagnostics or if you need help recreating.  I will be delighted to work with you live to open up screen shares or capture docs ....

kolban1@kolban.com

**Comment by vormplus** [comment]
Some of my students have the same problem, but it can take up to 5 minutes for sketches to start.
Working with the regular `size( 200, 200 )` is ok, adding the **P2D** or **P3D** renderer makes startup real slow.

**Comment by codeanticode** [comment]
That's odd, what's the hardware/OS you are using?

**Comment by vormplus** [comment]
One student has a Lenovo Y50, and one has a HP Pavilion. Both running Windows 10.
Processing 2.x seems to work fine, problem only exists with Processing 3.

**Comment by nkolban** [comment]
It seems that the common thread is Windows 10.  As mentioned previously, I will be happy to do a screen share to show the failing issue on a Windows 10 environment.

**Comment by codeanticode** [comment]
I have a windows 10 machine, will run some tests and will move from there. We can do the screen share if needed to.

**Comment by nkolban** [comment]
I installed 3.0.2 and there has been no change to the reported issue.  On my Windows 10 environment with an Nvidia graphics card and dual displays, it takes about 1 to 2 minutes for the "Primitives3D" example to open.

**Comment by lwlt** [comment]
I've seen this behaviour using the Processing IDE with the bundled JDK and also running as a standalone Java app with JDK 1.8.0u91 to make it easier to get to the debugger. The Java source I've been using is:

```
import processing.core.PApplet;
import processing.core.PImage;
import processing.core.PVector;



public class Main extends PApplet {

    public void settings() {
        size(400, 400, P2D);
        noSmooth();
    }


    static public void main(String[] passedArgs) {
        System.out.println("+++main()\n");
        String[] appletArgs = new String[] { "Main" };
        if (passedArgs != null) {
            PApplet.main(concat(appletArgs, passedArgs));
        } else {
            PApplet.main(appletArgs);
        }
    }

    @Override
    public void setup() {
        System.out.println("+++setup()");
        background(127);
    }

    @Override
    public void draw() {
        System.out.println("+++draw()");
        rect(100, 100, 200, 200);
        saveFrame("testgl-######.png");
        noLoop();
    }
}
```

There is a burst of CPU activity when the JVM starts, enters the main method, then the CPU idles while the sketch waits for 30-90 seconds not apparently doing anything. Due to the lack of CPU activity it seems like it's blocked on I/O rather than doing any computation. setup() doesn't run until after this delay. While waiting, the stack traces of each thread (via `where all` in jdb) are as follows: 

## AWT-Windows:
```
  [1] sun.awt.windows.WToolkit.eventLoop (native method)
  [2] sun.awt.windows.WToolkit.run (null)
  [3] java.lang.Thread.run (null)
```
## Java2D Disposer:
```
  [1] java.lang.Object.wait (native method)
  [2] java.lang.ref.ReferenceQueue.remove (null)
  [3] java.lang.ref.ReferenceQueue.remove (null)
  [4] sun.java2d.Disposer.run (null)
  [5] java.lang.Thread.run (null)
```
## Attach Listener:
## Signal Dispatcher:
## Finalizer:
```
  [1] java.lang.Object.wait (native method)
  [2] java.lang.ref.ReferenceQueue.remove (null)
  [3] java.lang.ref.ReferenceQueue.remove (null)
  [4] java.lang.ref.Finalizer$FinalizerThread.run (null)
```
## Reference Handler:
```
  [1] java.lang.Object.wait (native method)
  [2] java.lang.Object.wait (null)
  [3] java.lang.ref.Reference.tryHandlePending (null)
  [4] java.lang.ref.Reference$ReferenceHandler.run (null)
```
## main:
```
  [1] java.lang.ProcessImpl.waitForInterruptibly (native method)
  [2] java.lang.ProcessImpl.waitFor (null)
  [3] com.jogamp.common.util.IOUtil.testDirExec (IOUtil.java:1,009)
  [4] com.jogamp.common.util.IOUtil.testDirImpl (IOUtil.java:1,058)
  [5] com.jogamp.common.util.IOUtil.getSubTempDir (IOUtil.java:1,118)
  [6] com.jogamp.common.util.IOUtil.getTempDir (IOUtil.java:1,233)
  [7] com.jogamp.common.util.cache.TempFileCache.<clinit> (TempFileCache.java:81)
  [8] com.jogamp.common.util.cache.TempJarCache.initSingleton (TempJarCache.java:88)
  [9] com.jogamp.common.os.Platform$1.run (Platform.java:309)
  [10] java.security.AccessController.doPrivileged (native method)
  [11] com.jogamp.common.os.Platform.<clinit> (Platform.java:287)
  [12] com.jogamp.nativewindow.NativeWindowFactory$2.run (NativeWindowFactory.java:183)
  [13] java.security.AccessController.doPrivileged (native method)
  [14] com.jogamp.nativewindow.NativeWindowFactory.<clinit> (NativeWindowFactory.java:180)
  [15] com.jogamp.newt.NewtFactory$1.run (NewtFactory.java:68)
  [16] java.security.AccessController.doPrivileged (native method)
  [17] com.jogamp.newt.NewtFactory.<clinit> (NewtFactory.java:65)
  [18] processing.opengl.PSurfaceJOGL.initIcons (PSurfaceJOGL.java:531)
  [19] processing.opengl.PSurfaceJOGL.initFrame (PSurfaceJOGL.java:135)
  [20] processing.core.PApplet.initSurface (PApplet.java:10,374)
  [21] processing.core.PApplet.runSketch (PApplet.java:10,261)
  [22] processing.core.PApplet.main (PApplet.java:9,996)
  [23] Main.main (Main.java:16)
```  

Machine: Surfce 3, Atom x7-Z8700, Intel HD graphics, driver 2015-08-24, version 10.18.15.4248. Windows 10 version 1511 Build 10586.218.

I see the same behaviour with both OpenGL renderers, P2D and P3D. FX2D and the default renderer are fine.

**Comment by LujunWeng** [comment]
Same problem. It takes a long time to start processing 3.

**Comment by nkolban** [comment]
Its coming up for 6 months since I opened this issue report.  Is there anything further I can do to assist?  I am able to recreate at will and am willing to run any tests or provide live demonstration over skype (skype id: neil.kolban)

**Comment by codeanticode** [comment]
Hi, I was busy with other things over the last weeks, but will see if I can reproduce on my windows 10 system, and get back to you soon.

**Comment by nkolban** [comment]
No problems and thanks for your attention.  If it would help, I will be delighted to demonstrate on my Windows 10 system via Skype and allow you (supervised) access to run any tests you may need in case it is not easily re-creatable on your system.  I have noted that other users have also posted here so the good news is that, even if rare, it is less likely to be something "screwy" just to my machine.

**Comment by IggyZuk** [comment]
Is there a fix for this?

**Comment by nkolban** [comment]
Is it happening to you?

**Comment by IggyZuk** [comment]
Actually seems fine now, after my PC restarted and updated.

**Comment by bruce55** [comment]
Same here... Windows 10 14393 X64. It hangs in Processing IDE and IntelliJ IDEA.

**Comment by marzinp** [comment]
Hi, I have the same problem since Processing 3... Any fix available plz?

**Comment by MichaelHemingway** [comment]
So what's the version of Windows that doesn't have this issue? @IggyZuk 
I'm on 14393.447 as @bruce55 above and it's the same story with the hanging renderer.

**Comment by benfry** [comment]
Try disabling your virus scanner and see if that fixes things. Not a long-term solution, obviously, but the it looks like the virus scanner is checking the OpenGL support files that come with the sketch.

**Comment by diwi** [comment]
Just spent an afternoon to figure out why I lately have such a delay at program-start for OpenGL sketches (15-30 sec).
Disabling my virus scanner solved it too.

thx!

**Comment by gsim** [comment]
Alas stopping the anti-virus (Windows Defender) did not work for me, it made no difference. Processing takes about 15 to 20 minutes to open, and then it hangs.

**Comment by rschildge** [comment]
Has there been any resolution to this problem with Processing, Windows 10 and P3D? I have a 64 bit OS and the 64 bit Java JDK. The window eventually appears after a long delay.

**Comment by marzinp** [comment]
The problem disappeared here, maybe when I updated to 3.3.7, but I'm not sure...

**Comment by rschildge** [comment]
I tried 3.3.6 and 3.3.7 and they both have the same problem. Earlier versions work though.

**Comment by OctagonHex** [comment]
I just want to link to bug #4783, which is the underlying problem, at least for me.

**Comment by notd5a** [comment]
Im still getting this issue. P3D loads quite slowly and then the animation itself is really slow, doesnt run well in 3.5.4. I have an RTX 2070S and it still really slow.

**Comment by benfry** [comment]
Resolved for 4.0 beta 4 with https://github.com/processing/processing4/commit/be38429ef53b6108dff70855efaaadcab47fdd90
